# Bike Share Data Exploration

For the first part, I decided to use Jupyter notebooks. It is my first time working with this dataset, therefore I need to get familiar with the data before starting any analysis. 
Jupyter Notebook are a great tool for data exploration and visualization. However, I would prefer not to use them when building production systems.

## 1. Setup and Imports

In [1]:
# Import packages
import pandas as pd

## 2. Loading the Trip Data

I will use Pandas to read the trip data file into a DataFrame and start exploring.

In [2]:
trip_data_2020_10 = pd.read_csv('../bikeshare_data/202010-citibike-tripdata.csv')

I will visualize the first five rows in the data frame with the `head()` method.

In [3]:
trip_data_2020_10.head()

tripduration                 starttime                  stoptime  \
0          1403  2020-10-01 00:00:10.7870  2020-10-01 00:23:34.5400   
1           462  2020-10-01 00:00:18.2390  2020-10-01 00:08:00.5710   
2           770  2020-10-01 00:00:21.3070  2020-10-01 00:13:12.0840   
3           233  2020-10-01 00:00:25.0080  2020-10-01 00:04:18.6410   
4           126  2020-10-01 00:00:25.3100  2020-10-01 00:02:32.1330   

   start station id        start station name  start station latitude  \
0               224     Spruce St & Nassau St               40.711464   
1              3161    W 76 St & Columbus Ave               40.780184   
2               354   Emerson Pl & Myrtle Ave               40.693631   
3              3141           1 Ave & E 68 St               40.765005   
4               335  Washington Pl & Broadway               40.729039   

   start station longitude  end station id           end station name  \
0               -74.005524            3347  Van Brunt St & Wolcott St   
1               -73.977285            3158         W 63 St & Broadway   
2               -73.962236            3761      Cedar St & Myrtle Ave   
3               -73.958185             464            E 56 St & 3 Ave   
4               -73.994046             229             Great Jones St   

   end station latitude  end station longitude  bikeid    usertype  \
0             40.677343             -74.012751   39079  Subscriber   
1             40.771639             -73.982614   41267  Subscriber   
2             40.697842             -73.926241   47398    Customer   
3             40.759345             -73.967597   36482  Subscriber   
4             40.727434             -73.993790   45319  Subscriber   

   birth year  gender  
0        1995       1  
1        1998       2  
2        1996       1  
3        1993       1  
4        1989       1

## 3. Exploring the Trip Data

I will use the method `info()` to get more information about the columns.

In [4]:
trip_data_2020_10.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2248869 entries, 0 to 2248868
Data columns (total 15 columns):
 #   Column                   Dtype  
---  ------                   -----  
 0   tripduration             int64  
 1   starttime                object 
 2   stoptime                 object 
 3   start station id         int64  
 4   start station name       object 
 5   start station latitude   float64
 6   start station longitude  float64
 7   end station id           int64  
 8   end station name         object 
 9   end station latitude     float64
 10  end station longitude    float64
 11  bikeid                   int64  
 12  usertype                 object 
 13  birth year               int64  
 14  gender                   int64  
dtypes: float64(4), int64(6), object(5)
memory usage: 257.4+ MB


I can see that we have over two million entries in this particular month.

Also from this table, I can see that the **starttime and stoptime** are loaded with the data-type object. To use these components I need to convert them to timestamps.

To get a better feel for the numeric values I will use the `describe()` method.

In [5]:
trip_data_2020_10.describe().round(2)

tripduration  start station id  start station latitude  \
count    2248869.00        2248869.00              2248869.00   
mean        1213.05           2115.54                   40.74   
std        12014.57           1555.76                    0.04   
min           61.00             72.00                   40.66   
25%          421.00            435.00                   40.72   
50%          753.00           3115.00                   40.74   
75%         1316.00           3525.00                   40.77   
max      2245701.00           4230.00                   40.85   

       start station longitude  end station id  end station latitude  \
count               2248869.00      2248869.00            2248869.00   
mean                    -73.98         2112.33                 40.74   
std                       0.02         1556.50                  0.04   
min                     -74.03           72.00                 40.66   
25%                     -73.99          434.00                 40.72   
50%                     -73.98         3112.00                 40.74   
75%                     -73.96         3524.00                 40.77   
max                     -73.88         4230.00                 40.85   

       end station longitude      bikeid  birth year      gender  
count             2248869.00  2248869.00  2248869.00  2248869.00  
mean                  -73.98    37585.39     1981.75        1.18  
std                     0.02     9145.88       12.42        0.61  
min                   -74.07    14529.00     1873.00        0.00  
25%                   -73.99    32748.00     1970.00        1.00  
50%                   -73.98    39443.00     1985.00        1.00  
75%                   -73.96    45276.00     1992.00        2.00  
max                   -73.88    48703.00     2004.00        2.00

The **trip duration** is interesting because I can see that the shortest trips were 61 seconds and the longest was 2245701 seconds. These numbers are very spread out and even unrealistic, it seems like we have some outliers. I will need to address that.

Also the **birth year** seems to have some outliers because the min is 1873. The age of the person would be 148 years. 

I need to address these issues before making any useful analysis. However, before I do that I will use the `from_records()` method to get additional information about the data like the number of unique values, its data type and the amount of memory each column uses.

In [6]:
pd.DataFrame.from_records([(col, trip_data_2020_10[col].count(), trip_data_2020_10[col].nunique(), trip_data_2020_10[col].dtype, trip_data_2020_10[col].memory_usage(deep=True))
                           for col in trip_data_2020_10.columns], columns=['Column Name', 'Count', 'Unique', 'Data Type','Memory Usage'])

Column Name    Count   Unique Data Type  Memory Usage
0              tripduration  2248869    14364     int64      17991080
1                 starttime  2248869  2247222    object     182158517
2                  stoptime  2248869  2247175    object     182158517
3          start station id  2248869     1144     int64      17991080
4        start station name  2248869     1144    object     173041452
5    start station latitude  2248869     1144   float64      17991080
6   start station longitude  2248869     1142   float64      17991080
7            end station id  2248869     1163     int64      17991080
8          end station name  2248869     1163    object     173072676
9      end station latitude  2248869     1163   float64      17991080
10    end station longitude  2248869     1161   float64      17991080
11                   bikeid  2248869    18472     int64      17991080
12                 usertype  2248869        2    object     149580843
13               birth year  2248869      107     int64      17991080
14                   gender  2248869        3     int64      17991080

The values for the count are the same for each column, which means that there are no missing values.

## 4. Data Clean-up

The first thing I can do is convert the **start and end time** columns **from object to timestamp**.

In [7]:
trip_data_2020_10['starttime'] = pd.to_datetime(trip_data_2020_10['starttime'])
trip_data_2020_10['stoptime'] = pd.to_datetime(trip_data_2020_10['stoptime'])

Second, there are several columns that can be converted to categorical values. For example, the **start and end station names** are good candidates to be stored as **categorical data** (There are only 1144 unique values). 

The same applies to the **usertype** because it only has 2 unique values. **Gender** with three values and **bikeid** could also be converted to **categorical data**.

In [8]:
cols = ['start station name', 'end station name', 'bikeid', 'usertype', 'gender']
for col in cols:
    trip_data_2020_10[col] = trip_data_2020_10[col].astype('category')

If I run again the same analysis we can see that now the data types are more align with the content of the columns and also the memory usage was reduced with becomes very important when processing millions of data points.

In [9]:
pd.DataFrame.from_records([(col, trip_data_2020_10[col].count(), trip_data_2020_10[col].nunique(), trip_data_2020_10[col].dtype, trip_data_2020_10[col].memory_usage(deep=True))
                           for col in trip_data_2020_10.columns], columns=['Column Name', 'Count', 'Unique', 'Data Type','Memory Usage'])

Column Name    Count   Unique       Data Type  Memory Usage
0              tripduration  2248869    14364           int64      17991080
1                 starttime  2248869  2247222  datetime64[ns]      17991080
2                  stoptime  2248869  2247175  datetime64[ns]      17991080
3          start station id  2248869     1144           int64      17991080
4        start station name  2248869     1144        category       4620507
5    start station latitude  2248869     1144         float64      17991080
6   start station longitude  2248869     1142         float64      17991080
7            end station id  2248869     1163           int64      17991080
8          end station name  2248869     1163        category       4621835
9      end station latitude  2248869     1163         float64      17991080
10    end station longitude  2248869     1161         float64      17991080
11                   bikeid  2248869    18472        category       5174066
12                 usertype  2248869        2        category       2249237
13               birth year  2248869      107           int64      17991080
14                   gender  2248869        3        category       2249129

One last thing I could do before moving on to the next part is to set the minimum age for the birth column. I am afraid of modifying the other columns since it requires domain knowledge to confidently do feature engineering.

In [10]:
min_birth_year = 1950
trip_data_2020_10['birth year'] = trip_data_2020_10['birth year'].where(trip_data_2020_10['birth year'] >= min_birth_year, min_birth_year)

In [11]:
trip_data_2020_10.describe().round(2)

tripduration  start station id  start station latitude  \
count    2248869.00        2248869.00              2248869.00   
mean        1213.05           2115.54                   40.74   
std        12014.57           1555.76                    0.04   
min           61.00             72.00                   40.66   
25%          421.00            435.00                   40.72   
50%          753.00           3115.00                   40.74   
75%         1316.00           3525.00                   40.77   
max      2245701.00           4230.00                   40.85   

       start station longitude  end station id  end station latitude  \
count               2248869.00      2248869.00            2248869.00   
mean                    -73.98         2112.33                 40.74   
std                       0.02         1556.50                  0.04   
min                     -74.03           72.00                 40.66   
25%                     -73.99          434.00                 40.72   
50%                     -73.98         3112.00                 40.74   
75%                     -73.96         3524.00                 40.77   
max                     -73.88         4230.00                 40.85   

       end station longitude  birth year  
count             2248869.00  2248869.00  
mean                  -73.98     1981.82  
std                     0.02       12.18  
min                   -74.07     1950.00  
25%                   -73.99     1970.00  
50%                   -73.98     1985.00  
75%                   -73.96     1992.00  
max                   -73.88     2004.00

For example the **tripduration** sounds very unrealistic because it translates to 623 hours. Probably someone forgot to lock the bike.

## 5. Save intermediate results

For now I will save the intermediate results in Parquet format.

In [12]:
trip_data_2020_10.to_parquet('../intermediate_results/202010-citibike-tripdata-clean.parquet')